lib/proformaxml/services/transform_task.rb

```diff
@@ -23,26 +23,56 @@ def perform
     private
 
     def transform_from_2_0_to_2_1
-      if @task.submission_restrictions.present?
-        @task.submission_restrictions['submission-restrictions']['file-restriction'].each do |fr|
-          fr['@use'] = fr.delete('@required') == 'true' ? 'required' : 'optional'
-        end
-      end
+      transform_submission_restrictions_from_2_0_to_2_1 unless @task.submission_restrictions.nil?
+      transform_external_resources_from_2_0_to_2_1 unless @task.external_resources.nil?
 
       @task.model_solutions.filter! {|model_solution| model_solution.id != 'ms-placeholder' }
     end
 
-    def transform_from_2_1_to_2_0
-      unless @task.submission_restrictions.nil?
-        @task.submission_restrictions['submission-restrictions']['file-restriction'].each do |fr|
-          fr['@required'] = (fr.delete('@use') == 'required').to_s
-        end
-        @task.submission_restrictions['submission-restrictions'].delete('description')
-        @task.submission_restrictions['submission-restrictions'].delete('internal-description')
+    def transform_external_resources_from_2_0_to_2_1
+      ensure_array(@task.external_resources['external-resources']['external-resource']).each do |external_resource|
+        external_resource['@used-by-grader'] = external_resource['@used-by-grader'] || 'false'
+        external_resource['@visible'] = external_resource['@visible'] || 'no'
+      end
+    end
+
+    def transform_submission_restrictions_from_2_0_to_2_1
+      ensure_array(@task.submission_restrictions['submission-restrictions']['file-restriction']).each do |fr|
+        fr['@use'] = fr['@required'].nil? || fr.delete('@required') == 'true' ? 'required' : 'optional'
```

Comment:
Can you quickly explain this transformation?

If `@required` is present, the second part of the "or" is evaluated, so I understand that it's either 'required' or 'optional'.
If it is not present, i.e. `nil` here, it's set to `true` (Boolean)?

Reply:
If `@required` is not present, the condition of the ternary expression should be true and therefore evaluate to `'required'`. Since the precedence of `||` is higher than `?, :` it should be evaluated in the correct order. [1]
This is also covered by a test-case (`transform_task_spec.r:61`).

A small experiment also supports this approach:
```ruby
0> true || false ? 'a' : 'b'
=> a
```

[1](https://ruby-doc.org/core-2.6.2/doc/syntax/precedence_rdoc.html#label-Precedence)

Reply:
Despite being optional, we might want to add explicit parentheses.

Reply:
Sure.
fun fact: `or` has a lower precedence and would lead to `true`:
```ruby
0> true or false ? 'a': 'b'
=> true

```

Reply:
Rubocop does not agree:
```
lib/proformaxml/services/transform_task.rb:41:22: C: [Correctable] Style/TernaryParentheses: Omit parentheses for ternary conditions.
        fr['@use'] = (fr['@required'].nil? || fr.delete('@required') == 'true') ? 'required' : 'optional'
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Ouf, yes, I misunderstood the conditions. Parentheses would help. 

Does it work with `if ... else ... end`? It can also be used in variable assignments and may add some clarity.

Reply:
Sure, we  can do that

Reply:
Thanks! That's better, I'd say. Now, CodeCov also found another line not covered 🤓, which will decrease overall code coverage from 100% 🙀 No big deal, but if you have a few more minutes, another test case for that might be valuable.

Reply:
No big deal? 🙀

Reply:
Thanks, we're back at the optimum 💯

Reply:
@Dome-GER do you want to comment on this change again, or shall we proceed merging?